Replace makd with dub and upgrade deps

mihails-strasuns opened the pull request:
Major build system rework:

- each generated binary defined as separate DUB sub-package
- each top-level utlity library package is also defined as DUB sub-package
- tests redefined using dub configurations
- dependency versions upgraded to latest available (i.e. vibe.d 0.8)
- makd only used for packaging
- Travis script updated to match our current standards

This is one huge commit but all these changes are so subtly related that I really don't want to spend time trying to untangle this :(

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -11,56 +11,46 @@ services:
     global:
         - DIST=xenial
-        - ARTIFACTS_DIR=build/travis-artifacts
+        # Default beaver image names. May be overriden in specific stages
```
Are the changes to travis really that hard to split to a separate commit?

mihails-strasuns replied:
If you don't mind that it ends up with CI-failing intermediate commits - no :)

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -1,58 +1,33 @@
-$O/libvibed.a: $(shell find $C/submodules/vibed/source -type f -name '*.d')
-	$(call exec, dmd $(DFLAGS) -lib -of$@ $?)
+$B/neptune-autopr:
+	dub build :autopr
```
You can probably write an implicit rule, something like:

```make
tools := overview release autopr

all += $(patsubst %,$B/neptune-%,$(tools))

# Recompile on flag changes
$B/%: $G/build-d-flags

# Build with dub
$B/neptune-%: $C/src/%/main.d
    dub build :$*
```

mihails-strasuns replied:
Will try. Btw I don't use `.d` files as dependency intentionally here - dub does dependency tracking on its own (so running `dub build` again will be no-op) and it is much more convoluted than what I can possibly implement in a makefile.

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -1,58 +1,33 @@
+dub-unittest:
+	dub test :overview
+	dub test :release
+	dub test :autopr
```
Similarly:

```make
$O/unittest-%.stamp: $C/src/%/main.d
    dub test :$*
.PHONY: dub-unittest
dub-unittest: $(patsubst %,$O/unittest-%,$(tools))
```

But I think you could even redefine the `allunittest` target or something.

mihails-strasuns replied:
I tried that, but it looks like makd tries to do some automatic machinery despite that which fails (allunittests.d generation). Maybe not looked hard enough.

leandro-lucarella-sociomantic replied:
I don't know exactly what you tried, but in theory this should work because there is no `$O/unittest-%.stamp` defined in Makd. Maybe it could have failed for `$O/test-%.stamp` because that's already defined so they might compete. If you change that to `$O/inttest-%.stamp` for now I think it should work too.

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -1,58 +1,33 @@
+	dub run :tests -c first_release   -- $(ITFLAGS)
+	dub run :tests -c prerelease      -- $(ITFLAGS)
+	dub run :tests -c patch_release   -- $(ITFLAGS)
+	dub run :tests -c preview_release -- $(ITFLAGS)
```
And same again:

```make
$O/test-%.stamp: $C/$(INTEGRATIONTEST)/%/main.d
    dub run :tests -c $*   -- $(ITFLAGS)

.PHONY: dub-inttest
dub-inttest: $(patsubst $(INTEGRATIONTEST)/%/main.d,$O/test-%.stamp,$(wildcard $(INTEGRATIONTEST)/*/main.d))
```

(all this untested, of course, but just so you get the idea :) )

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -1,58 +1,33 @@
-autopr: $B/neptune-autopr
-
-all += release overview autopr
+dub-test: dub-unittest dub-inttest
```
Why the `dub-*`instead of the regular makd targets?

mihails-strasuns replied:
They are defined in makd with `:=` :(

leandro-lucarella-sociomantic commented on Makefile:
```diff
@@ -5,3 +5,6 @@ include submodules/makd/Makd.mak
 
+# DUB doesn't work well with flavors but there isn't any need in those
+# for neptune apps anyway:
+VALID_FLAVORS := dub
```
ah... So you want to make both available concurrently? :thinking:

mihails-strasuns replied:
Unless I am missing something obvious, dub doesn't support different output directories per build config (and it defines about 10 standard build presets, not just debug and release). As we don't care that much for neptune which one is used I decided to just use the latest (they overwrite each other).

leandro-lucarella-sociomantic commented on src/internal/git/helper.d:
```diff
@@ -10,9 +10,9 @@
 *******************************************************************************/
 
-module lib.git.helper;
+module internal.git.helper;
```
Why the name change from `lib` to `internal`?

mihails-strasuns replied:
To make it more obvious that this is not user-facing library package and is not exposed via `subPackage` intentionally. I plan to give a try to refactoring/restructuring this all a bit after initial PR in an attempt to make every top-level package "public".

leandro-lucarella-sociomantic replied:
This also really looks like it should go in a separate commit, it shouldn't be that hard.

leandro-lucarella-sociomantic commented on src/octod/api/authorizations.d:
```diff
@@ -0,0 +1,58 @@
+
+*******************************************************************************/
+
+module octod.api.authorizations;
```
What? You are also moving octod here in this same commit? :scream:

mihails-strasuns replied:
Yup, this is not strictly a prerequisite, but makes things much more simple dub.sdl definition-wise.

I will split it into separate one.

leandro-lucarella-sociomantic reviewed (changes requested):
Changes look goodish to me in general, I think if we move forward in this direction eventually we could even create a `MakDub.mak` file in `makd` to use the same conventions as in `Makd.mak` but with these adaptions to dub.

What I think is a bit unacceptable is the mess of one commit to rule them all. I can understand some parts might be too interrelated and hard to split, but I think at least the octod part (which is even undocumented in the PR description) and the travis changes should be split. I think even if we ended up with some commit that doesn't build it would be better than this huge commit.

mihails-strasuns commented:
@leandro-lucarella-sociomantic 

>  I think if we move forward in this direction eventually we could even create a MakDub.mak file in makd to use the same conventions as in Makd.mak but with these adaptions to dub.

I think that at the very least there is a solid case to split packaging related bits of makd into own `MakPkg.mak` so that it can be used in any project (even non-D) without any other MakD conventions coming into effect.

Commit:
Move octod source to the neptune repo

Because of plans to both move to GitLab and transition to using GrapQL
API where applicable there is a little value in octod as a separate
project anymore and it is moved in-place as legacy dependency.

Commit:
Rename `lib` package to `internal`

Makes it more obvious that it is collection of internal utilities
not intended for public usage. Intention is to eventually look into
those and reorganize into one or more user-facing library packages.

Commit:
Update .travis.yml to match latest standards

Copied from dhtnode and adjusted to neptune specifics

Commit:
Replace makd with dub and upgrade deps

Major build system rework:

- each generated binary defined as separate DUB sub-package
- each top-level utlity library package is also defined as DUB sub-package
- dependency versions upgraded to latest available (i.e. vibe.d 0.8)
- makd only used for packaging and wrapping dub calls

mihails-strasuns commented:
Updated:

- split into several commits (low hanging fruits)
- more generalized makefile definition
- `dub -q` calls (quite, warnings/errors only) to make travis log more readable

mihails-strasuns-sociomantic commented:
Ping @leandro-lucarella-sociomantic

leandro-lucarella-sociomantic reviewed (approved):
Thanks, looks much better. Now I don't get why the rename `integrationtest` -> `tests`, but other than that LGTM.

mathias-baumann-sociomantic commented:
I checked out the branch and tried `make test` with the following result:


```
 marenz@labs-081  ~/sociomantic/neptune   pr-223  make test                                                                     :-) ~/sociomantic/neptune 
   gen build/dub/tmp/allunittests.d
   rdmd build/dub/tmp/allunittests
./src/octod/core.d(68): Error: no identifier for declarator Connection
./src/octod/core.d(68): Error: alias cannot have initializer
./src/octod/core.d(103): Error: found 'import' instead of statement
./src/octod/core.d(124): Error: found 'import' instead of statement
./src/octod/core.d(131): Error: found '!' when expecting ';' following statement
./src/octod/core.d(142): Error: found 'import' instead of statement
./src/octod/core.d(143): Error: found '!' when expecting ';' following statement
./src/octod/core.d(143): Error: found '..' when expecting ']'
./src/octod/core.d(143): Error: found '$' when expecting ')'
./src/octod/core.d(143): Error: semicolon expected to close declaration, not ']'
./src/octod/core.d(143): Error: found ']' instead of statement
./src/octod/core.d(201): Error: basic type expected, not scope
./src/octod/core.d(201): Error: found 'scope' when expecting ')'
./src/octod/core.d(201): Error: semicolon expected following function declaration
./src/octod/core.d(201): Error: found 'request' when expecting ',' following argument
./src/octod/core.d(201): Error: expression expected, not ')'
./src/octod/core.d(201): Error: found '{' when expecting ',' following argument
./src/octod/core.d(202): Error: found ';' when expecting ',' following argument
./src/octod/core.d(203): Error: found ';' when expecting ',' following argument
./src/octod/core.d(204): Error: found ';' when expecting ',' following argument
./src/octod/core.d(205): Error: expression expected, not '}'
Failed: ["dmd1", "-di", "-g", "-I/home/marenz/sociomantic/neptune/build/dub/include", "-I./src", "-I./submodules/makd/src", "-I./submodules/beaver/src", "-unittest", "-debug=UnitTest", "-version=UnitTest", "-v", "-o-", "/home/marenz/sociomantic/neptune/build/dub/tmp/allunittests.d", "-I/home/marenz/sociomantic/neptune/build/dub/tmp"]
submodules/makd/Makd.mak:595: recipe for target '/home/marenz/sociomantic/neptune/build/dub/tmp/allunittests' failed
make: *** [/home/marenz/sociomantic/neptune/build/dub/tmp/allunittests] Error 1
```

mathias-baumann-sociomantic commented:
Using `DMD64 D Compiler v2.081.1`

mihails-strasuns commented:
`make test` won't work because of current limitations in `makd` (I can't disable all automatic targets it implies). Because of that new `make dub-test` was added for CI purpose.

mihails-strasuns commented:
Thanks!
